core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1941,6 +1941,16 @@ class Log(@volatile var dir: File,
     }
   }
 
+  @threadsafe
+  private[log] def getFirstBatchTimestampForSegments(segments: Iterable[LogSegment]): Iterable[Long] = {
```

Comment:
One Optimization we can have here is to lock only when we need to populate firsBatchTimestamp, for example:
in LogSegment.scala , we can add:
def isFirstBatchTimestampLoaded : Boolean = {
    rollingBasedTimestamp match {
      case Some(t) => true
      case _ => false
    }
  }

In log.scala, we can have:

@threadsafe
  private[log] def getFirstBatchTimestampForSegments2(segments: Iterable[LogSegment]): Iterable[Long] = {

    val needLoadTimestampSegments = segments.filter {
      segment =>
        !segment.isFirstBatchTimestampLoaded
    }

    if (needLoadTimestampSegments.nonEmpty) {
      lock synchronized {
        needLoadTimestampSegments.foreach {
          segment =>
            segment.loadFirstBatchTimestamp()
        }
      }
    }

    segments.map {
      segment =>
        segment.getFirstBatchTimestampWithoutLoading()
    }
  }

This optimization will reduce lock contention on the log.  However, each time we call log.logSegments,  we are acquiring the locks.  log.logSegments has been called multiple times (e.g, cleanableOffsets, LogToClean, etc.) during the log compaction . 

The choice here is :  reduced potential lock contention versus simplified code.

Reply:
This isn't exactly introduced by your patch, but I'm not very clear on the thread-safety semantics offered by the log lock. E.g., in `getEstimatedEarliestTime` the first line grabs log segments using the underlying log lock but then relinquishes it.

That said, do you even need to lock? What is the worst that can happen? e.g., this could result in a failed attempt to read the earliest batch timestamp of a segment that may have been deleted after listing all the segments but I don't think this is actually an issue.

Reply:
This is really good point.  I don't see lock is hold when logcleaner performs the actual compaction IOs : cleanSegments -> cleanInto.   Maybe we can skip the lock. Let me double check.

Reply:
can you add a comment here explaining why this is safe? per your explanation offline, segments cannot be removed outside of the cleaner (for log compacted topics).